liquidctl/driver/coolit.py

```diff
@@ -0,0 +1,415 @@
+"""liquidctl driver for Corsair Platinum and PRO XT coolers.
+
+Supported devices
+-----------------
+
+ - Corsair H110i GT
+
+Supported features
+------------------
+
+ - general monitoring
+ - pump speed control
+ - fan speed control
+
+Copyright Roberto Marques, Serphentas, and other contributors
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+import logging
+import re
+
+from enum import Enum, unique
+
+from liquidctl.driver.usb import UsbHidDriver
+from liquidctl.keyval import RuntimeStorage
+from liquidctl.util import clamp, fraction_of_byte, u16le_from, normalize_profile
+
+
+LOGGER = logging.getLogger(__name__)
+
+_REPORT_LENGTH = 64
+_PUMP_INDEX = 0x02
+
+_COMMAND_FIRMWARE_ID = 0x01
+_COMMAND_TEMP_READ = 0x0E
+_COMMAND_FAN_SELECT = 0x10
+_COMMAND_FAN_MODE = 0x12
+_COMMAND_FAN_FIXED_PWM = 0x13
+_COMMAND_FAN_FIXED_RPM = 0x14
+_COMMAND_FAN_READ_RPM = 0x16
+_COMMAND_FAN_MAX_RPM = 0x17
+_COMMAND_FAN_RPM_TABLE = 0x19
+_COMMAND_FAN_TEMP_TABLE = 0x1A
+
+_OP_CODE_WRITE_ONE_BYTE = 0x06
+_OP_CODE_READ_ONE_BYTE = 0x07
+_OP_CODE_WRITE_TWO_BYTES = 0x08
+_OP_CODE_READ_TWO_BYTES = 0x09
+_OP_CODE_WRITE_THREE_BYTES = 0x0A
+_OP_CODE_READ_THREE_BYTES = 0x0B
+
+_PROFILE_LENGTH = 5
+_CRITICAL_TEMPERATURE = 60
+
+_PUMP_DEFAULT_QUIET = [0x2E, 0x09]
+_PUMP_DEFAULT_EXTREME = [0x86, 0x0B]
+
+_SEQUENCE_MIN = 1
+_SEQUENCE_MAX = 255
+
+
+@unique
+class _FanMode(Enum):
+    FIXED_DUTY = 0x02
+    FIXED_RPM = 0x04
+    CUSTOM_PROFILE = 0x0E
+
+    @classmethod
+    def _missing_(cls, value):
+        LOGGER.debug("falling back to FIXED_DUTY for _FanMode(%s)", value)
+        return _FanMode.FIXED_DUTY
+
+
+@unique
+class _PumpMode(Enum):
+    QUIET = 0x08
+    EXTREME = 0x0C
+
+    @classmethod
+    def _missing_(cls, value):
+        LOGGER.debug("falling back to QUIET for _PumpMode(%s)", value)
+        return _PumpMode.QUIET
+
+
+def _sequence():
+    """Return a generator that produces valid protocol sequence numbers.
+
+    Sequence numbers are random.
+    """
+    num = 1
+    while True:
+        yield (num % 31) + 1
```

Comment:
Is this really needed for this device? In the protocol docs you said a simple counter is sufficient, and this is slightly more complex that that.

Reply:
The fact that random sequence numbers didn't work well suggests that some jumps in sequence numbers aren't acceptable to the device. Thus restarting the sequence at every invocation should also sometimes cause issues, and persisting the sequence in `storage` may be necessary after all.

While numbers modulo 32 seem to work, why not using the simpler full byte range (i.e. modulo 256), as seen in the captures? Well, actually, the full range except `0`, which doesn't appear to be used. Something like

```py
yield (num % 255) + 1
```

Reply:
We could use that but it implies testing the device again. Perhaps a quicker method (that's known to work) is to use the [old implementation](https://github.com/liquidctl/liquidctl/pull/147/files#diff-96821345dcd62e3b3b04ad2e571f1e2e53d79f4de38446c9ab40ef9f298f3b3cR91).

Reply:
Fair point. And since this appears to work, and from sanmai's comment there that in general any per-command sequence seems to work, let's just keep things as is. At least for now/until bugs are reported...